File: src/domain/__tests__/labels.ts
import { Labels } from '~/domain/labels';

const appNameTest = (strLabels: string[], expectedAppName: string | null) => {
  const labels = strLabels.map(l => Labels.toKV(l));
  const appName = Labels.findAppNameInLabels(labels);

  expect(appName).toBe(expectedAppName);
};

describe('extract app name', () => {
  test('from reserved host', () => {
    appNameTest(['reserved:host'], 'host');
    appNameTest(['reserved:host='], 'host');
  });

  test('from reserved world', () => {
    appNameTest(['reserved:world'], 'world');
    appNameTest(['reserved:world='], 'world');
  });

  test('from reserved health', () => {
    appNameTest(['reserved:health'], 'health');
    appNameTest(['reserved:health='], 'health');
  });

  test('from reserved init', () => {
    appNameTest(['reserved:init'], 'init');
    appNameTest(['reserved:init='], 'init');
  });

  test('from reserved remote-node', () => {
    appNameTest(['reserved:remote-node'], 'remote-node');
    appNameTest(['reserved:remote-node='], 'remote-node');
  });

  test('from reserved kube-apiserver', () => {
    appNameTest(['reserved:kube-apiserver'], 'kube-apiserver');
    appNameTest(['reserved:kube-apiserver='], 'kube-apiserver');
  });

  test('from reserved unmanaged', () => {
    appNameTest(['reserved:unmanaged'], 'unmanaged');
    appNameTest(['reserved:unmanaged='], 'unmanaged');
  });

  test('from reserved unknown', () => {
    appNameTest(['reserved:unknown'], 'unknown');
    appNameTest(['reserved:unknown='], 'unknown');
  });

  test('from named cidr group (cilium/hubble-ui#1051)', () => {
    const cg = 'cidrgroup:io.cilium.policy.cidrgroupname/klab-dns-west';
    appNameTest([cg, 'reserved:world'], 'klab-dns-west');
    appNameTest(['reserved:world', cg], 'klab-dns-west');
    appNameTest(['reserved:world', 'cidrgroup:role=klab-dns'], 'world');
  });

  test('cidr group icon label', () => {
    const labels = ['reserved:world', 'cidrgroup:icon=etcd'].map(l => Labels.toKV(l));
    expect(Labels.findCIDRGroupIconInLabels(labels)).toBe('etcd');
    expect(Labels.detect(labels).cidrGroupIcon).toBe('etcd');
  });

  test('cidr group site becomes the cluster badge', () => {
    const labels = ['reserved:world', 'cidrgroup:site=central'].map(l => Labels.toKV(l));
    expect(Labels.detect(labels).clusterName).toBe('central');
    // a real cluster label wins over the group site
    const both = [
      'reserved:world',
      'k8s:io.cilium.k8s.policy.cluster=west',
      'cidrgroup:site=central',
    ].map(l => Labels.toKV(l));
    expect(Labels.detect(both).clusterName).toBe('west');
  });

  test('from k8s prefixed lbls', () => {
    appNameTest(['k8s:app=EXPECTED_NAME'], 'EXPECTED_NAME');
    appNameTest(['k8s:name=EXPECTED_NAME'], 'EXPECTED_NAME');
    appNameTest(['k8s:functionName=EXPECTED_NAME'], 'EXPECTED_NAME');
    appNameTest(['k8s:k8s-app=EXPECTED_NAME'], 'EXPECTED_NAME');

    appNameTest(['k8s:io.kubernetes.pod.app=EXPECTED_NAME'], 'EXPECTED_NAME');
    appNameTest(['k8s:io.kubernetes.pod.name=EXPECTED_NAME'], 'EXPECTED_NAME');
    appNameTest(['k8s:io.kubernetes.pod.functionName=EXPECTED_NAME'], 'EXPECTED_NAME');
    appNameTest(['k8s:io.kubernetes.pod.k8s-app=EXPECTED_NAME'], 'EXPECTED_NAME');

    appNameTest(['app=EXPECTED_NAME'], 'EXPECTED_NAME');
    appNameTest(['name=EXPECTED_NAME'], 'EXPECTED_NAME');
    appNameTest(['functionName=EXPECTED_NAME'], 'EXPECTED_NAME');
    appNameTest(['k8s-app=EXPECTED_NAME'], 'EXPECTED_NAME');

    appNameTest(['k8s:app.kubernetes.io/name=EXPECTED_NAME'], 'EXPECTED_NAME');
  });
});

File: backend/domain/labels/labels.go
package labels

import (
	"strings"
)

type ReservedLabel string
type SpecialLabel string

const (
	ReservedLabelHost          ReservedLabel = "reserved:host"
	ReservedLabelWorld         ReservedLabel = "reserved:world"
	ReservedLabelHealth        ReservedLabel = "reserved:health"
	ReservedLabelInit          ReservedLabel = "reserved:init"
	ReservedLabelRemoteNode    ReservedLabel = "reserved:remote-node"
	ReservedLabelUnmanaged     ReservedLabel = "reserved:unmanaged"
	ReservedLabelUnknown       ReservedLabel = "reserved:unknown"
	ReservedLabelKubeAPIServer ReservedLabel = "reserved:kube-apiserver"

	SpecialLabelKubeDNS    SpecialLabel = "k8s:k8s-app=kube-dns"
	SpecialLabelPrometheus SpecialLabel = "k8s:app=prometheus"
)

var (
	prefixes = []string{"k8s:", "io.kubernetes.pod.", "app.kubernetes.io/"}
	appKeys  = []string{"app", "name", "functionName", "k8s-app"}
)

type LabelProps struct {
	IsHost          bool
	IsWorld         bool
	IsRemoteNode    bool
	IsKubeDNS       bool
	IsKubeAPIServer bool
	IsInit          bool
	IsHealth        bool
	IsPrometheus    bool
	AppName         *string
	// CIDRGroupName is the name of the CiliumCIDRGroup covering a world
	// identity (cilium >= 1.17 stamps it as a key-only label). It is what
	// lets the UI keep one service card PER named group instead of the
	// single aggregated world card.
	CIDRGroupName *string
}

// CIDRGroupNameLabelPrefix is matched against the RAW label: `cidrgroup:` is
// not among the normalized prefixes, and the group name (key tail) must be
// preserved as-is.
const CIDRGroupNameLabelPrefix = "cidrgroup:io.cilium.policy.cidrgroupname/"

func Props(labels []string) *LabelProps {
	props := new(LabelProps)

	for _, lbl := range labels {
		k, v := LabelAsKeyValue(lbl, true)

		props.IsHost = props.IsHost || k == string(ReservedLabelHost)
		props.IsWorld = props.IsWorld || k == string(ReservedLabelWorld)
		props.IsInit = props.IsInit || k == string(ReservedLabelInit)
		props.IsHealth = props.IsHealth || k == string(ReservedLabelHealth)
		props.IsRemoteNode = props.IsRemoteNode || k == string(
			ReservedLabelRemoteNode,
		)
		props.IsKubeDNS = props.IsKubeDNS || lbl == string(SpecialLabelKubeDNS)
		props.IsPrometheus = props.IsPrometheus || k == string(
			SpecialLabelPrometheus,
		)
		props.IsKubeAPIServer = props.IsKubeAPIServer || k == string(
			ReservedLabelKubeAPIServer,
		)

		if props.CIDRGroupName == nil {
			if name, ok := strings.CutPrefix(lbl, CIDRGroupNameLabelPrefix); ok {
				name, _, _ = strings.Cut(name, "=")
				if len(name) > 0 {
					props.CIDRGroupName = &name
				}
			}
		}

		if props.AppName != nil {
			continue
		}

		props.AppName = appNameFromKV(k, v)
	}

	return props
}

func LabelAsKeyValue(lbl string, normalize bool) (string, string) {
	parts := strings.SplitN(lbl, "=", 2)
	n := len(parts)

	k, v := "", ""
	if n >= 1 {
		k = parts[0]
	}

	if n == 2 {
		v = parts[1]
	}

	if normalize {
		k = NormalizeKey(k)
	}

	return k, v
}

func NormalizeKey(labelKey string) string {
	for _, prefix := range prefixes {
		labelKey = strings.ReplaceAll(labelKey, prefix, "")
	}

	return labelKey
}

func appNameFromKV(k, v string) *string {
	for _, appKey := range appKeys {
		if k == appKey {
			return &v
		}
	}

	return nil
}
